crates/bevy_render/src/render_resource/gpu_list.rs

```diff
@@ -0,0 +1,127 @@
+use super::StorageBuffer;
+use crate::{
+    render_resource::batched_uniform_buffer::BatchedUniformBuffer,
+    renderer::{RenderDevice, RenderQueue},
+};
+use bevy_ecs::{prelude::Component, system::Resource};
+use encase::{private::WriteInto, ShaderSize, ShaderType};
+use std::{marker::PhantomData, mem};
+use wgpu::{BindGroupLayoutEntry, BindingResource, BindingType, BufferBindingType, ShaderStages};
+
+/// Trait for types able to go in a [`GpuList`].
+pub trait GpuListable: ShaderType + ShaderSize + WriteInto + Clone {}
+impl<T: ShaderType + ShaderSize + WriteInto + Clone> GpuListable for T {}
+
+/// Stores a list of elements to be transferred to the GPU and made accessible to shaders as a read-only array.
+///
+/// On platforms that support storage buffers, this is equivalent to [`StorageBuffer<Vec<T>>`].
+/// Otherwise, this falls back to batched uniforms.
+///
+/// Other options for storing GPU-accessible data are:
+/// * [`StorageBuffer`](crate::render_resource::StorageBuffer)
+/// * [`DynamicStorageBuffer`](crate::render_resource::DynamicStorageBuffer)
+/// * [`UniformBuffer`](crate::render_resource::UniformBuffer)
+/// * [`DynamicUniformBuffer`](crate::render_resource::DynamicUniformBuffer)
+/// * [`GpuList`](crate::render_resource::GpuList)
+/// * [`BufferVec`](crate::render_resource::BufferVec)
+/// * [`Texture`](crate::render_resource::Texture)
+#[derive(Resource)]
+pub enum GpuList<T: GpuListable> {
+    Uniform(BatchedUniformBuffer<T>),
+    Storage((StorageBuffer<Vec<T>>, Vec<T>)),
+}
+
+impl<T: GpuListable> GpuList<T> {
+    pub fn new(device: &RenderDevice) -> Self {
+        let limits = device.limits();
+        if limits.max_storage_buffers_per_shader_stage < 3 {
```

Comment:
this 3 maybe originated in the shader cache `NO_STORAGE_BUFFERS` test, where i think it was chosen because the light bindings needed 3 storage buffers. as written it'll always be true or always false, so that if a card supports 8 (the default) but a pipeline creates 10 `GpuList`s, it'll fall over.

ideally this would use the actual number of storage buffers created so far, so that it could use storage buffers up to the actual limit before falling back to uniforms (and shader code could query appropriately using the same limit). i was thinking about passing an index in (or maybe a `<const I: usize>` generic parameter) to allow users to say what the count so far is for the current pipeline/stage. it's not perfect since the caller would need to know all the GpuList bindings upfront (across view/material/mesh etc, e.g.)... but better than just using 3 i think.

not sure, but i think this needs addressing somehow.

Reply:
I'm not really sure whether bevy should bother trying to check resource usage (like storage buffer bindings per stage or whatever the limit is) because wgpu anyway checks it and points it out. I suppose we could add some validation into our `RenderDevice` wrapper when creating bind group layouts and/or bind groups. The pros of us also doing validation would be that we could perhaps provide what we think are more useful error messages. The cons... well, it's more code in bevy that has to be maintained. :) Thoughts?

Reply:
i'm not saying bevy should track it automatically in any scenario, but pipeline authors could do so.

as it is it pipeline authors using `GpuList` will have to choose a minimum spec to support, and use only that many `GpuList` bindings, and fall back to dynamic uniforms for more. then they would still support platforms with no storage buffers, but not platforms with x storage buffers where 3 <= x < what they use.

i don't know if storage buffers per stage is a limit that varies much or is ever low ... if not then no big deal. if so then it sounds fragile (if author chooses to use a lot) or limiting (if they choose to use fewer than they could).

i guess it doesn't matter much for meshes since you're going to use precisely one storage buffer ... certainly happy to kick it down the road to avoid blocking that.

Reply:
It's a good point. And good idea about making it so that one can perhaps prioritise bindings that should be storage buffers, and then overflow to uniform buffers. That said, I think it should be handled in a separate PR as it is probably quite an intrusive change to make and needs more thought to make it work well and conveniently.